src/v/kafka/server/handlers/fetch.cc

```diff
@@ -702,6 +711,186 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+    }
+
+    std::chrono::milliseconds current_delay(const ss::scheduling_group& sg) {
+        auto sampled_time = ss::sched_clock::now();
+        auto sample = sg.get_stats().runtime;
+        auto total_busy_time = ss::engine().total_busy_time();
+
+        auto dt = sampled_time - _last_sampled_time;
+        if (dt < sample_rate) {
+            return _last_delay;
+        }
+
+        auto busy_de = total_busy_time - _last_total_busy_time;
+        if (busy_de < 0 * duration_unit) {
+            return _last_delay;
+        }
+        int64_t busy = (scaling_factor * busy_de) / dt;
+        if (busy > scaling_factor) {
+            return _last_delay;
+        }
+
+        auto de = sample - _last_sample;
+        if (de < 0 * duration_unit) {
+            return _last_delay;
+        }
+        int64_t current_runtime = (scaling_factor * de) / dt;
+        if (current_runtime > scaling_factor) {
+            return _last_delay;
+        }
+
+        // If reactor utilization ~ 100% then ensure the fetch runtime is
+        // less than or equal to _target_fetch_sg_util.
+        //
+        // Otherwise ensure that reactor utilization is 100% even if the
+        // fetch runtime is above _target_fetch_sg_util.
+        constexpr int64_t target_reactor_util = 9990;
+        int64_t error = (busy < target_reactor_util)
+                          ? (busy - target_reactor_util)
+                          : (current_runtime - _target_fetch_sg_util);
+
+        // calculate the proportional
+        int64_t p = error / _p_coeff;
+
+        // calculate the integral
+        _error_int += error * (dt / duration_unit);
+        // Letting the integral fall below 0 only reduces the responsiveness
+        // in cases of sudden load spikes.
+        _error_int = std::max(0L, _error_int);
+        int64_t i = _error_int / _i_coeff;
+
+        // calculate the derivative
+        int64_t d = (error - _last_error) / (dt / duration_unit);
+        d *= _d_coeff;
+
+        auto pid = p + i + d;
+        pid = std::max(0L, pid);
+
+        auto delay = std::chrono::duration_cast<std::chrono::milliseconds>(
+          duration_unit * pid);
+        delay = std::min(delay, _max_delay);
+
+        _last_delay = delay;
+        _last_error = error;
+        _last_sample = sample;
+        _last_total_busy_time = total_busy_time;
+        _last_sampled_time = sampled_time;
+        _error_var(error);
+
+        return delay;
+    }
+
+private:
+    void reset() {
+        _last_delay = 0ms;
+        _error_int = 0;
+        _last_error = 0;
+        _last_sample = 0ms;
+        _last_total_busy_time = 0ms;
+        _last_sampled_time = ss::sched_clock::now();
+        _error_var = {};
+    }
+
+    void setup_metrics() {
+        namespace sm = ss::metrics;
+        if (config::shard_local_cfg().disable_metrics()) {
+            return;
+        }
+
+        _metrics.add_group(
+          prometheus_sanitize::metrics_name("kafka_fetch_pid"),
```

Comment:
Isn't the most important metric a gauge of how much debounce is being applied at the moment or perhaps the average debounce over some period? It seems like that's the thing you'd check first: "Is dynamic debounce active, and what is it doing?".

Reply:
I didn't include debounce directly since its a dependent variable that can be calculated from the other metrics. Maybe it would be easier to include it directly though.

Reply:
Did you have a Grafana formula for that? I seems like to get close it would require basically reproducing all the logic in `current_delay` including knowing all the configuration values, etc, and compensating for the fact you have only 1 minute (or whatever) samples for things like fetch group usage.

Reply:
Good point, will add a counter metric `delay_seconds_total` to avoid the need for a complicated prom query.